statsrelay/src/backend.rs

```diff
@@ -1,33 +1,39 @@
 use std::collections::HashMap;
 use std::sync::atomic::AtomicU64;
+use std::sync::Arc;
 
 use futures::stream::FuturesUnordered;
 use regex::bytes::RegexSet;
 
+use crate::backend_client::prom_client::PromClient;
+use crate::backend_client::statsd_client::StatsdClient;
+use crate::backend_client::BackendClient;
 use crate::config;
+use crate::config::BackendClientConfig;
 use crate::discovery;
 use crate::shard::{statsrelay_compat_hash, Ring};
 use crate::stats;
-use crate::statsd_client::StatsdClient;
 use crate::statsd_proto;
 use crate::statsd_proto::Event;
 
 use log::warn;
 
-pub struct StatsdBackend {
-    conf: config::StatsdBackendConfig,
-    ring: Ring<StatsdClient>,
+type ArcBackendClient = Arc<dyn BackendClient + Send + Sync>;
+
+pub struct Backend {
+    conf: config::BackendConfig,
+    ring: Ring<ArcBackendClient>,
     input_filter: Option<RegexSet>,
     warning_log: AtomicU64,
     backend_sends: stats::Counter,
     backend_fails: stats::Counter,
 }
 
-impl StatsdBackend {
+impl Backend {
```

Comment:
I'm not _certain_ there's a cleaner factoring available here, but it does seem like this type has a couple responsibilities that might be reasonably layered/composed rather than belonging to one type -- constructing the correct backend from a given config, handling discovery updates, managing a consistent hashing ring of shards, tracking filters, and carrying around stats.

I know the stats just kind of have to be passed down, so that's not really separable, but I wonder if maybe the config -> backend bit could be pulled out to a free function this calls, and/or perhaps composed in a ShardedBackendClient type to manage the ring, leaving the layering something like `Backend(filters, stats, config, ShardedBackendClient(config, stats, ring, BackendClient(config)))`?

It's not immediately clear to me whether or not this is worth doing at the current stage, but I do think if we're expecting to further extend this type it runs the risk of becoming an all-in-one class with a lot of branching logic that's more difficult to test and compose. Layering it would hopefully make that outcome less likely.